smugup.1
.\"     Title: smugup
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.71.0 <http://docbook.sf.net/>
.\"      Date: May 2008
.\"    Manual: smugup
.\"    Source: smugup
.\"
.TH "SMUGUP" "1" "May 2008" "smugup" "smugup"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
smugup \- upload files to a smugmug.com gallery
.SH "SYNOPSIS"
.HP 7
\fBsmugup\fR [\fB\-\-email\ email@address\fR] [\fB\-\-password\ password\fR] [\-\-album\ album\ [\-\-category\ category]\ [\-\-settings\ quicksettings]] [\fB\-\-quiet\fR] [\fB\-\-debug\fR] [\fB\-\-version\fR] [\fB\-\-help\fR]
.SH "DESCRIPTION"
.PP
smugup uploads files to a specified smugmug.com gallery. You can specify multiple files to upload on the command line. If the email address or password is not specified, you will be prompted for it. If the album name is not specified, you will be promted with a list of the album names in this account and you can choose one from there.
.PP
smugup supports both images and videos to be uploaded, as long as your smugmug.com account allows them.
.SH "OPTIONS"
.PP
\fB\-\-email email@address\fR
.RS 3n
Specify the email address of your smugmug.com account.
.RE
.PP
\fB\-\-password password\fR
.RS 3n
Specify the password of your smugmug.com account.
.RE
.PP
\fB\-\-album album\fR
.RS 3n
Specify the target album. If ommited, the user will be prompted with a list. If the album and category are both specified, the album will be created if necessary.
.RE
.PP
\fB\-\-category category\fR
.RS 3n
Specify the target category for
\fIalbum creation\fR. (See
\fB\-\-album\fR.)
.RE
.PP
\fB\-\-settings quicksettings\fR
.RS 3n
Specify the target quicksettings for
\fIalbum creation\fR. (See
\fB\-\-album\fR.)
.RE
.PP
\fB\-\-debug\fR
.RS 3n
Print a whole bunch of debugging messages to stdout.
.RE
.PP
\fB\-\-quiet\fR
.RS 3n
Run as quietly as possible, not printing anything to stdout.
.RE
.PP
\fB\-\-version\fR
.RS 3n
Print version number.
.RE
.PP
\fB\-\-help\fR
.RS 3n
Print help text.
.RE
.SH "AUTHOR"
.PP
Written by Greg Kroah\-Hartman
<greg@kroah.com>.
.SH "SEE ALSO"
.PP
\fBsmugls\fR(1),
\fBsmugdown\fR(1)


ChangeLog
Summary of changes from v004 to v005
============================================

troy (3):
      smugup enhanced to create the Album if necessary,
      smugup enhanced to create the Album if necessary.
      Another extraneous file.

Summary of changes from v003 to v004
============================================

Andreas Jaeger (6):
      Free memory leaks
      smugbatch.spec: Prepare for 004
      Add upload/download speed in KiB/s.
      Error handling for login:
      Add select_album function
      Use select_album

Greg Kroah-Hartman (21):
      move to 004 for Makefile to not get confused
      add smugls.3 man file
      fix section in smugls.xml
      add smugup.3
      add --version to smugls
      some coding style fixups
      man pages should be in 1, not 3
      fix comment in list.h to make a bit more sense
      add scripts/checkpatch.pl
      list.h - checkpatch.pl fixups
      smug.h - checkpatch.pl fixups
      Merge branch 'master' of /home/greg/src/smugbatch//
      smug_core.c - checkpatch.pl fixups
      Merge branch 'master' of /home/greg/src/smugbatch//
      smug_core.c - fix up coding style issues
      add HACKING file to help explain the coding style and how to check for it.
      added README file
      fix memory leak in select_album()
      fix memory leak in smug_login()
      add valgrind info to HACKING
      document some of the calls in smug_core.c to show

Summary of changes from v002 to v003
============================================

Andreas Jaeger (7):
      zalloc can be a macro
      Add a spec file
      Allow CFLAGS as argument to make
      Handle fgets return value
      Add config parsing
      Prepare Spec file for version 3
      Fix curl warning

Greg Kroah-Hartman (21):
      added zalloc for zeroed out malloc to make things easier
      save off the original url of an image so we can download it if needed
      smugup:  add --quiet option
      s/which/Which/ when asking for an album name
      add a header to the list of files and handle filenames with no name set
      add file list to album and document struct album
      put files that are in an album in the album instead of the session wide list
      add smugdown.c (for downloading files)
      add smugdown to .gitignore
      smugdown: look at the proper list of files in the album
      Add -O2 to CFLAGS
      free the password and email if we have read it from the config file for smugup and smugdown
      smug_core: fix bug with trailing ;
      smug_core: coding style cleanups to make checkpatch happy
      delete old smugbatch.c file
      smugdown: fix coding style issues found by checkpatch
      smugls: fix coding style issues found by checkpatch
      smugup: fix coding style issues found by checkpatch
      make the buffer really start at 0
      smugdown: start to work on dowloading, don't have the correct url just yet...
      fix up some compiler warnings in smug_core


Summary of changes from v001 to v002
============================================

Andreas Jaeger (2):
      Always set the ssl sanity checks
      Typo fix for smugup and smugls

Greg Kroah-Hartman (7):
      fix problem of not setting the filename for the images uploaded
      increment version number...
      added smugls to .gitignore
      remove compiler warnings from smugls
      build both smugup and smugls
      add ability to read album contents to smug_core.c
      make smugls actually list the files...


Summary of changes from v000 to v001
============================================

Greg KH (2):
      make download show numbers in int not double and use only basename
      remove some of the libraies we need

Greg Kroah-Hartman (45):
      added api key
      added smugup-1.1.sh as the framework to work off of
      .gitignore added
      can read the session id properly
      added simplessl.c example curl code
      clean up code a bit, add other needed urls
      add license and option parsing
      add COPYING
      Makefile rewrite.
      fix compiler warning, hiding the fact that we were not reporting the error properly..
      properly read the buffer sent to us by curl
      now we can log out
      set the proper useragent for the curl session
      don't null out the session_id when reading the logout data
      send the request for the list of albums
      start of filename parsing...
      add smugbatch_version.h to .gitignore
      automatically generate smugbatch_version.h
      switch to the 1.2.0 version of the smugmug api
      we can now read albums
      added --debug
      added md5.c and md5.h from coreutils 6.9 release.
      add md5 to build
      get generation of md5 working properly
      added *.swp to .gitignore
      upload works!
      remove simplessl as it's not needed
      added test image to play with
      add progress function
      properly handle the basename of the uploaded file
      add *.jpg to .gitignore
      make debug global in smugbatch.c
      add smug.h and smug_core.c
      add smug_login and smug_logout
      add album function
      add session_alloc() and session_free()
      added smugup.c
      fix memory problem in smug_core.c
      smugup: make the check for login properly...
      Makefile: switch to smugup instead of smugdump
      add smugup to .gitignore
      add help to smugup
      add ability to choose album to upload to
      add version to help
      add release target to Makefile



smugup.c
/*
 * Copyright (C) 2008 Greg Kroah-Hartman <greg@kroah.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/stat.h>
#include "list.h"
#include "md5.h"
#include "smugbatch_version.h"
#include "smug.h"


int debug;

static void display_help(void)
{
	fprintf(stdout, "smugup - upload photos to smugmug.com\n");
	fprintf(stdout, "Version: " SMUGBATCH_VERSION "\n");
	fprintf(stdout, "Usage:\n");
	fprintf(stdout, "  smugup [options] files\n");
	fprintf(stdout, "options are:\n");
	fprintf(stdout, "  --email email@address\n");
	fprintf(stdout, "  --password password\n");
	fprintf(stdout, "  --album album [ --category category"
				" --settings quicksettings ]\n");
	fprintf(stdout, "  --quiet\n");
	fprintf(stdout, "  --debug\n");
	fprintf(stdout, "  --help\n");
}

int main(int argc, char *argv[], char *envp[])
{
	static const struct option options[] = {
		{ "debug", 0, NULL, 'd' },
		{ "email", 1, NULL, 'e' },
		{ "password", 1, NULL, 'p' },
		{ "album", 1, NULL, 'a' },
		{ "category", 1, NULL, 'c' },
		{ "settings", 1, NULL, 's' },
		{ "help", 0, NULL, 'h' },
		{ "quiet", 0, NULL, 'q' },
		{ }
	};
	struct filename *filename;
	struct session *session;
	struct album *album;
	char *album_title = NULL;
	char *category_title = NULL;
	char *quicksettings_name = NULL;
	int retval;
	int option;
	int i;

	session = session_alloc();
	if (!session) {
		fprintf(stderr, "no more memory...\n");
		return -1;
	}

	curl_global_init(CURL_GLOBAL_ALL);
	smug_parse_configfile(session);

	while (1) {
		option = getopt_long_only(argc, argv, "dqe:p:a:c:s:h",
					  options, NULL);
		if (option == -1)
			break;
		switch (option) {
		case 'd':
			debug = 1;
			break;
		case 'e':
			if (session->email)
				free(session->email);
			session->email = strdup(optarg);
			dbg("email = %s\n", session->email);
			break;
		case 'p':
			if (session->password)
				free(session->password);
			session->password = strdup(optarg);
			dbg("password = %s\n", session->password);
			break;
		case 'a':
			album_title = strdup(optarg);
			dbg("album_title = %s\n", album_title);
			break;
		case 'c':
			category_title = strdup(optarg);
			dbg("category_title = %s\n", category_title);
			break;
		case 's':
			quicksettings_name = strdup(optarg);
			dbg("quicksettings_name = %s\n", quicksettings_name);
			break;
		case 'q':
			session->quiet = 1;
			break;
		case 'h':
			display_help();
			goto exit;
		default:
			display_help();
			goto exit;
		}
	}

	if ((category_title || quicksettings_name) && !album_title) {
		fprintf(stdout, "Category and Settings requires Title.");
		display_help();
		goto exit;
	}

	if (!session->email) {
		fprintf(stdout, "Enter smugmug.com email address: ");
		session->email = get_string_from_stdin();
	}

	if (!session->password) {
		fprintf(stdout, "Enter smugmug.com password: ");
		session->password = get_string_from_stdin();
	}


	/* build up a list of all filenames to be used here */
	for (i = optind; i < argc; ++i) {
		filename = zalloc(sizeof(*filename));
		if (!filename)
			return -ENOMEM;
		filename->filename = strdup(argv[i]);
		filename->basename = my_basename(filename->filename);
		dbg("adding filename '%s'\n", argv[i]);
		list_add_tail(&filename->entry, &session->files_upload);
	}

	retval = smug_login(session);
	if (retval) {
		fprintf(stderr, "Can not login\n");
		return -1;
	}

	dbg("1\n");
	retval = smug_get_albums(session);
	if (retval) {
		fprintf(stderr, "Can not read albums\n");
		return -1;
	}

	retval = generate_md5s(&session->files_upload);
	if (retval) {
		fprintf(stderr, "Error calculating md5s\n");
		return -1;
	}
	album = select_album(album_title, category_title, quicksettings_name,
			     session);
	if (!album)
		return -1;
	retval = upload_files(session, album);
	if (retval) {
		fprintf(stderr, "Error uploading files\n");
		return -1;
	}

	smug_logout(session);

	session_free(session);
exit:
	return 0;
}
